src/EditorFeatures/Core/ExternalAccess/EditorConfig/Features/Diagnostics/DiagnosticHandler.cs

```diff
@@ -0,0 +1,310 @@
+        }
+
+        private record struct DocumentDiagnosticSource(TextDocument Document) : IDiagnosticSource
+        {
+            public ProjectOrDocumentId GetId() => new(Document.Id);
+
+            public Project GetProject() => Document.Project;
+
+            public Uri GetUri() => Document.GetURI();
+
+            public async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(IDiagnosticAnalyzerService diagnosticAnalyzerService, RequestContext context, DiagnosticMode diagnosticMode, CancellationToken cancellationToken)
+            {
+                var document = context.AdditionalDocument;
+                Contract.ThrowIfNull(document);
+
+                var workspace = document.Project.Solution.Workspace;
+
+                var filePath = document.FilePath;
+                Contract.ThrowIfNull(filePath);
+
+                var optionSet = workspace.Options;
+                var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                var diagnostics = new List<DiagnosticData>();
+
+                var definedSettings = new HashSet<string>();
+
+                foreach (var line in text.Lines)
+                {
+                    var lineText = line.ToString();
+
+                    // Check that setting definition doesn't have more than 1 '='
+                    if (lineText.Where(c => c == '=').Count() > 1)
+                    {
+                        diagnostics.Add(CreateDiagnosticData(document, line, EditorConfigDiagnosticIds.IncorrectSettingDefinition, DiagnosticSeverity.Error, false, 1, Document.Project.Id));
+                    }
+                    else if (lineText.Contains('='))
+                    {
+                        var splitted = lineText.Split('=');
+                        string leftSide = splitted[0].Trim(), rightSide = splitted[1].Replace(" ", "");
+
+                        var settingsSnapshots = SettingsHelper.GetSettingsSnapshots(workspace, filePath);
+                        var settingsItems = settingsSnapshots.Select(sett => sett.GetSettingName()).WhereNotNull();
+
+                        // Check that the setting name exists
+                        if (ShowDiagnosticForSettingName(settingsItems, leftSide))
+                        {
+                            diagnostics.Add(CreateDiagnosticData(document, line, EditorConfigDiagnosticIds.SettingNotFound, DiagnosticSeverity.Error, false, 1, Document.Project.Id));
+                            continue;
+                        }
+
+                        // Check that the setting has not been defined
+                        if (definedSettings.Contains(leftSide))
+                        {
+                            diagnostics.Add(CreateDiagnosticData(document, line, EditorConfigDiagnosticIds.SettingAlreadyDefined, DiagnosticSeverity.Warning, false, 1, Document.Project.Id));
+                        }
+                        definedSettings.Add(leftSide);
```

Comment:
Unless I'm missing something, this should work?

```suggestion
                        if (!definedSettings.Add(leftSide))
                        {
                            diagnostics.Add(CreateDiagnosticData(document, line, EditorConfigDiagnosticIds.SettingAlreadyDefined, DiagnosticSeverity.Warning, false, 1, Document.Project.Id));
                        }
```